### DISCUSSION TITLE
Welcome to pyCoilGen Discussions!

### DISCUSSION THREAD

**Discussion by kev-m** [kev-m/pyCoilGen#28] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by szj7398** [comment]


**Reply by kev-m** [reply]
Hello! Thank you for your support.

I am not 100% sure that I fully understand your question. 
As described in the documentation, section [5.3 CoilSolution](https://pycoilgen.readthedocs.io/en/stable/results.html#coilsolution), the computed magnetic field is present in the `target_field` of the `CoilSolution` class.
You can get a hint on how to access the data in the [S2 Shim Coil code fragment](https://pycoilgen.readthedocs.io/en/stable/results.html#example-results-s2-shim-coil).

Does this help?

**Reply by szj7398** [reply]
Thank you for your answer. The issue I'm encountering is that the magnetic field data I input was collected on the surface of a sphere, with around 1500 sample points. However, after running the program, the number of coordinates stored in the `coilsolution` does not correspond to my input coordinates, which makes it impossible for me to calculate the homogeneity after shimming. Could it be that the method I’m using is problematic? Thank you again for your help.

**Reply by kev-m** [reply]
Can you create and share a project that shows what you are doing? I'd like to see your inputs, e.g. how you define the target field, and how you are reading the `CoilSolution` afterwards, to extract the calculated field.

**Comment by susanna-m-weber** [comment]
Hello,

I had a question about the way that the current strength for a coil is calculated by pyCoilGen and CoilGen. I am trying to simulate a coil that can generate a 100 mT homogenous field by setting the 'field_shape_function' parameter to 100. While generating the coil itself has worked well, the current strength determined by the program, which is generally on the scale of of 10^6 or 10^7 for the cylindrical coils I am simulating. Double checking with the paper, I can see that the current strengths reported for the shim coils are on the scale of 10^4, but other results, namely the brain shimming coil, show a current strength of less than 1 amp in the figures. The very large current values seem unreasonable to me, and I am also a bit confused by how the current strength is different from the 'opt_current_layout' variable, the ideal current strength for the connected layout to match the target field. 

Could you clarify why there is such a large range of current strengths shown in the paper figures, and what user errors could result in the program outputting such massive currents? 

Thank you very much for your help!

**Comment by kev-m** [comment]
Hello,
A quick observation: The top of the [configuration section](https://pycoilgen.readthedocs.io/en/stable/configuration.html) notes that all units must be in standard SI.

> Note: All values are specified in SI units, i.e. metre, Ampere, Tesla.

I infer that by setting the 'field_shape_function' value to 100, you are requesting a 100T field. 
This would probably justify the 10^6 A current that you are seeing.
Could this explain what you are experiencing?

**Reply by susanna-m-weber** [reply]
Hi,

Thank you very much for your reply! 

I assumed that the field strength was set in mT rather than in T because of the figure labeling produced by the “plot_errors” function - when I enter a value of 0.1, it shows the target Bz as being 0.1 mT on the figure produced. 

Using this lower field strength, CoilGen  outputs a current strength on the scale of 10^4 A, which still seems too large for a 100 mT field. How does this value differ from the "opt_current_layout" variable that both CoilGen and pyCoilGen output? 

Thank you again for your help!

**Reply by kev-m** [reply]
Hello!

> I assumed that the field strength was set in mT rather than in T because of the figure labeling produced by the “plot_errors” function - when I enter a value of 0.1, it shows the target Bz as being 0.1 mT on the figure produced.

That does sound like a bug in the figure label. The value should only be mT if the value is divided.

> Using this lower field strength, CoilGen outputs a current strength on the scale of 10^4 A, which still seems too large for a 100 mT field. 

This is best checked with another, independent, application. Are you aware of any other tool that can be used to cross-check this result?

> How does this value differ from the "opt_current_layout" variable that both CoilGen and pyCoilGen output?

Good question. All I know is the [line of code that computes this value](https://github.com/kev-m/pyCoilGen/blob/0.2.3/pyCoilGen/sub_functions/evaluate_field_errors.py#L161): 
`opt_current_layout = np.abs(np.mean(target_field.b[2, :] / combined_field_layout[2, :]))`

If I examine the the [TargetField data-structure](https://github.com/kev-m/pyCoilGen/blob/0.2.3/pyCoilGen/sub_functions/data_structures.py#L646), I believe that the `2` index means that it is sampling only the z-coordinate.

**Comment by szj7398** [comment]
Hello! When using PyCoilGen to design gradient coils, if I want to plot the linear variation of the x-direction gradient field within a cross-section (such as the xoy plane) rather than only observing the linear variation on the surface of a sphere, can PyCoilGen do this?

**Reply by kev-m** [reply]
Hello! All the [data](https://github.com/kev-m/pyCoilGen/blob/master/pyCoilGen/sub_functions/data_structures.py) is available in the coil solution, for you to examine and plot however you wish.
Have a look at the [plotting example](https://github.com/kev-m/pyCoilGen/blob/master/examples/plotting_s2_shim.py) and the [plotting functions](https://github.com/kev-m/pyCoilGen/tree/master/pyCoilGen/plotting), for ideas.
Does this help?

**Reply by szj7398** [reply]
I may not have expressed my meaning clearly. The `coords` in the `targetfield` input are coordinates on a spherical surface. What I want to do now is plot data for a custom cross-section, which is not stored in `coilsolution`. How can I upload the coordinates of a custom cross-section and then plot it?

**Reply by kev-m** [reply]
> I may not have expressed my meaning clearly. The `coords` in the `targetfield` input are coordinates on a spherical surface. What I want to do now is plot data for a custom cross-section, which is not stored in `coilsolution`. How can I upload the coordinates of a custom cross-section and then plot it?

Perhaps the plot like that shown in [5.5.1.3. The Target Field Z-Component](https://pycoilgen.readthedocs.io/en/stable/results.html#the-target-field-z-component) is closer to what you're describing?

Have a look at the details of the [plot_vector_field() function](https://github.com/kev-m/pyCoilGen/blob/master/pyCoilGen/plotting/plot_vector_field.py#L10C5-L10C22), specifically lines 31 through 37.

That function is extracting a plane of results from the [TargetField](https://github.com/kev-m/pyCoilGen/blob/master/pyCoilGen/sub_functions/data_structures.py#L642), which is a volume of points, not just a surface.
Actually, that depends on how your project is set up. If you do not specify a [target mesh](https://pycoilgen.readthedocs.io/en/stable/configuration.html#target-field), then the target field is a spherical volume.
